Add test for species transport model

ansjsia opened the pull request:
Test the following features of the species transport model:

- Turning on the species transport model
- Checking the mixtures object command list
- Creating a custom mixture with custom species (using the TUI)
- Copying an existing mixture
- Reassigning a cellzone material to a different mixture
- Deleting an existing mixture

Currently, the species transport model cannot be used without resorting to TUI commands due to the following open issues with the settings API:

- Creating/copying mixtures: https://github.com/ansys/pyfluent/issues/2275
- Listing mixtures, species, and properties: https://github.com/ansys/pyfluent/issues/2277
- Deleting a species from a mixture: https://github.com/ansys/pyfluent/issues/2259
- Creating a mixture from an empty dictionary: https://github.com/ansys/pyfluent/issues/2274

dnwillia-work commented on tests/test_solvermode/test_speciestransport.py:
```diff
@@ -0,0 +1,61 @@
+    # Test turning on species transport model
+    species_mdl = solver_session.setup.models.species.model
+
+    assert species_mdl.option.get_state() == "off"
```
Do you have to call `.get_state()` here?  I would think this should work:
`assert species_mdl.option == "off"`

raph-luc replied:
What does `species_mdl.option` return? Compare it to that instead of the string `"off"`

ansjsia replied:
`species_mdl.option` is of type `ansys.fluent.core.solver.settings_241.option.option`. Calling `species_mdl.option()` returns a `str`.

raph-luc replied:
Right, thanks for the details. And does `species_mdl.option()` return the same as `species_mdl.option.get_state()`?

You are correct that using `get_state()` is more explicit, but in my understanding PyFluent users should generally not need these more API-backend calls such as `get_state`, `get_attr` and so on. In other words, we do not really want the more explicit commands here, we want to showcase the more simplified/easier-to-use interfaces if possible

ansjsia replied:
Yes, just calling the object itself returns the same thing. I'll replace the `get_state()` call with just `()`.

dnwillia-work commented on tests/test_solvermode/test_speciestransport.py:
You should name the module `test_species_model.py`.  A bit shorter and add an underscore between the two words.

dnwillia-work commented on tests/test_solvermode/test_speciestransport.py:
```diff
@@ -0,0 +1,61 @@
+@pytest.mark.quick
+@pytest.mark.setup
+@pytest.mark.fluent_version("latest")
+def test_species_material(load_mixing_elbow_mesh):
```
Maybe name it `test_create_mixture`?

ansjsia replied:
What if we called it `test_change_create_mixture` since I'm also editing, copying, and deleting objects?

raph-luc replied:
I would say `test_species_materials` (added `s` at the end) is slightly better, as you are testing `setup.species` and `setup.materials` (`mixture` is under `materials` as well). 

Alternatively, if you would like to compete for the longest test names with Mainak, a very detailed `test_species_materials_mixture_create_change_copy_delete` would be great, as these are all the operations you are testing here

But any one of these suggested sound OK to me

dnwillia-work replied:
It's true that the test is asserting a number of things but for me the main point of it is to illustrate how to create a new mixture from scratch.  For me a longer name is a bit cluttered and `test_species_materials` does not seem accurate enough.

raph-luc replied:
I think that moving the creation part into a function and breaking this down into two or three separate tests might be a little excessive, but could work as well.

And I agree, I previously meant that the suggested `test_create_mixture` or `test_change_create_mixture` sound fine to me too

ansjsia commented:
@dnwillia-work @raph-luc For your consideration in the latest commit. I've renamed the file to `test_species_model.py` and split it up into multiple tests. I created a fixture such that only one solver instance would be used across all tests, ensuring that we only start and shut down Fluent once for all six tests.

The tests are:
1. Activate the species transport model
2. Check the command list
3. Create a custom mixture from custom species
4. Duplicate a mixture
5. Change the mixture assigned to a cellzone
6. Delete a mixture

The problem is that if one of these tests fails, all the following tests will fail as well, like dominoes. I'm not sure if this is acceptable. If it isn't, a potential solution is to implement recognizing an `incremental` test marker in our `conftests.py` file that would skip all remaining tests in the class if one fails: https://docs.pytest.org/en/latest/example/simple.html#incremental-testing-test-steps

raph-luc commented on tests/test_solvermode/test_speciestransport.py:
Make sure to remove this file from the branch, as it is no longer applicable

raph-luc commented on tests/test_solvermode/test_species_model.py:
```diff
@@ -0,0 +1,103 @@
+    solver_session.exit()
+
+
+@pytest.mark.quick
+@pytest.mark.setup
```
We are not using these markers currently, so IMO they are just unnecessary copy-pasted repetition. I would remove them:
```suggestion
```

dnwillia-work replied:
What do you think, should we also remove them from `pyproject.toml`?  That's where they are defined:

https://github.com/ansys/pyfluent/blob/53aa0270d1ab0d9bdc91c0b6d749297357656c8a/pyproject.toml#L16

raph-luc replied:
That would make sense to me, the only markers we are actually using in the GitHub CIs are: `nightly`, `fluent-version`, `solvermode`, `codegen_required` and `standalone`.

In my understanding, we could remove all of these as they aren't being used, and aren't providing any helpful info that the test name or module name don't already provide:
https://github.com/ansys/pyfluent/blob/53aa0270d1ab0d9bdc91c0b6d749297357656c8a/pyproject.toml#L17-L22

raph-luc replied:
Discussion in https://github.com/ansys/pyfluent/discussions/2287

dnwillia-work replied:
OK, so lets leave `setup` but remove everything else that are not used currently.

raph-luc commented on tests/test_solvermode/test_species_model.py:
```diff
@@ -0,0 +1,103 @@
+from ansys.fluent.core.examples import download_file
+
+
+@pytest.fixture(scope="module")
+def solver_session():
+    cas_path = download_file(
+        "mixing_elbow.cas.h5",
+        "pyfluent/mixing_elbow",
+    )
+    solver_session = pyfluent.launch_fluent(mode="solver")
+    solver_session.file.read_case(file_name=cas_path)
+    yield solver_session
+    solver_session.exit()
```
We have usually been specifying fixtures in separate files as they are re-used in multiple places, but as this one is specific to this test module, having it here makes sense to me too. 

Also, good that you are re-using one session for multiple tests, rather than launching a separate session for each test

ansjsia replied:
Yeah, I created a specific fixture for this because the original mixing elbow fixture is function-scoped, but I needed a module-scoped fixture to run multiple tests on the same solver object.

dnwillia-work commented on tests/test_solvermode/test_speciestransport.py:
I guess you don't need this file now?

dnwillia-work replied:
Yeah, the commits don't lie, looks like you did not.  Ideally what you would have done in this case is `git mv old_name.py new_name.py` then implemented additional changes into `new_name.py`.

raph-luc replied:
If you deleted it manually using File Explorer for example, you'd need to add the file delete command to the commit with `git add tests/test_solvermode/test_speciestransport.py`

Either moving it with `git mv` as Dan suggested, or removing it directly with `git rm` (which deletes the file and adds it to the commit) would also have worked

raph-luc commented on tests/test_solvermode/test_species_model.py:
```diff
@@ -0,0 +1,103 @@
+    assert elbow_zone.material() == "custom-mixture-copy"
+
+
+@pytest.mark.quick
+@pytest.mark.setup
+@pytest.mark.fluent_version(">=24.1")
+def test_delete_mixture(solver_session):
+    materials = solver_session.setup.materials
+    materials.mixture.delete(name_list=["custom-mixture"])
+    assert "custom-mixture" not in materials.mixture.keys()
```
The way these are structured now, these aren't "unit" tests, i.e. they aren't isolated. 

If I run `test_delete_mixture` on its own for example, it is always going to fail, and this is not how we've been structuring the tests. Here they only work if all the tests run, and if they run in the specific order defined in this file.

I think keeping everything as one test like we had before was ultimately better

dnwillia-work replied:
Yeah, agree, unit tests must work in isolation and should not have dependencies on other unit tests.

dnwillia-work commented on tests/test_solvermode/test_species_model.py:
```diff
@@ -0,0 +1,103 @@
+        "pyfluent/mixing_elbow",
+    )
+    solver_session = pyfluent.launch_fluent(mode="solver")
+    solver_session.file.read_case(file_name=cas_path)
```
May as well add the lightweight case read option here.

dnwillia-work replied:
Yeah, sorry, I mean case settings only mode:
https://ansyshelp.ansys.com/account/secured?returnurl=/Views/Secured/corp/v232/en/flu_ug/flu_ug_CaseSettingsOnly.html?q=case%20settings

raph-luc replied:
Ohhh, that makes perfect sense, thanks for the reference!

dnwillia-work replied:
yeah, I did not design this part of the API.  You should add an issue about your feedback :-).

dnwillia-work replied:
@ansjsia I guess this thread is not relevant now since you removed the fixture.  Maybe you can add a fixture that provides lightweight case read mode later.

ansjsia replied:
Absolutely, I'll make a separate PR for this.

ansjsia commented:
@raph-luc @dnwillia-work I've recombined everything into one test in the latest commit. Hopefully the name and comments are sufficiently descriptive.

dnwillia-work commented on tests/test_solvermode/test_species_model.py:
```diff
@@ -0,0 +1,72 @@
+
+
+@pytest.mark.setup
+@pytest.mark.fluent_version("latest")
```
You missed setting 24.1 or newer here.

ansjsia replied:
Oops, good catch. Fixed.

ansjsia commented:
@dnwillia-work @raph-luc The unit testing check for 23R2 is stalling (timing out after 6 hours) consistently on the test `test_meshing_workflow.py::test_watertight_workflow_children`, which seems completely unrelated to this PR. It seems to fail when I run it locally on 23R2. Any insight?

raph-luc commented:
@ansjsia I believe it wasn't `test_watertight_workflow_children` that stalled (as the logs say that it passed, and it should pass if you try it locally as well), it was the test after that one: `test_watertight_workflow_dynamic_interface`.

The team is aware of these tests that have been randomly stalling, we aren't sure of what is going on yet.

I set the failed test to re-run, let's see if it hangs again or if it passes. Changes in `main` branch are set to skip this test, we can merge them here if it fails again
